consensus/obcpbft/obc-batch.go

```diff
@@ -407,83 +394,78 @@ func (op *obcBatch) resubmitStaleRequest(c custodyInfo) {
+		}
+		return nil
+	case batchTimerEventID:
+		logger.Info("Replica %d batch timer expired", op.pbft.id)
+		if op.pbft.activeView && (len(op.batchStore) > 0) {
+			op.sendBatch()
+		}
+	case viewChangedEventID:
+		// Outstanding reqs doesn't make sense for batch, as all the requests in a batch may be processed
+		// in a different batch, but PBFT core can't see through the opaque structure to see this
+		// so, on view change, we rely on the fact that the complaint service will resubmit requests
+		// and instead zero the outstandingReqs map ourselves
+		op.pbft.outstandingReqs = make(map[string]*Request)
+
+		logger.Debug("Replica %d batch thread recognizing new view", op.pbft.id)
+		op.inViewChange = false
+		if op.batchTimerActive {
+			op.stopBatchTimer()
+		}
 
-			op.complainer.Restart()
-			for _, pair := range op.complainer.CustodyElements() {
-				logger.Info("Replica %d resubmitting request under custody: %s", op.pbft.id, pair.Hash)
-				op.submitToLeader(pair.Request)
-			}
-		case c := <-op.custodyTimerChan:
-			if !op.deduplicator.IsNew(c.req.(*Request)) {
-				op.resubmitStaleRequest(c)
-				continue
-			}
+		op.complainer.Restart()
+		for _, pair := range op.complainer.CustodyElements() {
+			logger.Info("Replica %d resubmitting request under custody: %s", op.pbft.id, pair.Hash)
+			op.submitToLeader(pair.Request)
+		}
+	case batchExecEventID:
+		execInfo := event.(batchExecEvent)
+		op.executeImpl(execInfo.seqNo, execInfo.raw)
+	case complaintEventID:
+		c := event.(complaintEvent)
+		logger.Debug("Replica %d processing complaint from custodian", op.pbft.id)
+		if !op.deduplicator.IsNew(c.req.(*Request)) {
+			op.resubmitStaleRequest(c)
+			break
+		}
 
-			if !c.complaint {
-				logger.Warning("Batch replica %d custody expired, complaining: %s", op.pbft.id, c.hash)
-				op.broadcastMsg(&BatchMessage{&BatchMessage_Complaint{c.req.(*Request)}})
+		if !c.complaint {
+			logger.Warning("Batch replica %d custody expired, complaining: %s", op.pbft.id, c.hash)
+			op.broadcastMsg(&BatchMessage{&BatchMessage_Complaint{c.req.(*Request)}})
+		} else {
+			if !op.inViewChange && op.pbft.activeView {
+				logger.Debug("Batch replica %d complaint timeout expired for %s", op.pbft.id, c.hash)
+				op.inViewChange = true
+				op.pbft.sendViewChange()
 			} else {
-				if !op.inViewChange && op.pbft.activeView {
-					logger.Debug("Batch replica %d complaint timeout expired for %s", op.pbft.id, c.hash)
-					op.inViewChange = true
-					op.pbft.injectChan <- func() { op.pbft.sendViewChange() }
-				} else {
-					logger.Debug("Batch replica %d complaint timeout expired for %s while in view change", op.pbft.id, c.hash)
-				}
+				logger.Debug("Batch replica %d complaint timeout expired for %s while in view change", op.pbft.id, c.hash)
 			}
-		case execInfo := <-op.execChan:
-			op.executeImpl(execInfo.seqNo, execInfo.raw)
-		case op.idleChan <- struct{}{}:
-			// Only used to detect thread idleness during unit tests
 		}
+	default:
```

Comment:
I think we should have two synchronous state machines (each with their own event queue), which are serviced by a single state machine manager, instead of doing this event pass-through.